hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/HeapData.java

```diff
@@ -39,17 +39,23 @@
 
     // array (12: array header, 4: length)
     private static final int ARRAY_HEADER_SIZE_IN_BYTES = 16;
-
-    protected byte[] payload;
+    protected ByteBuffer payload;
+    private final int length;
 
     public HeapData() {
+        length = 0;
     }
 
     public HeapData(byte[] payload) {
-        if (payload != null && payload.length > 0 && payload.length < HEAP_DATA_OVERHEAD) {
+        this(ByteBuffer.wrap(payload));
```

Comment:
Creating an additional `ByteBuffer` for each data object. I'm not sure about cost vs benefit.